[PW-4474] add refund notification processor

peterojo opened the pull request:
<!-- 🎉 Thank you for submitting a pull request! 🎉  -->

## Summary
<!--
Describe the changes proposed in this pull request:
- What is the motivation for this change? 
- What existing problem does this pull request solve?
-->
Create REFUND notification processor.
Transition order transactions to refunded/refunded_partially when notification is received and handle illegal state transitions
## Tested scenarios
<!-- Description of tested scenarios -->
open -> refunded (illegal: moves to paid and then to refunded) ✅
refunded -> refunded (illegal: skips) ❌
refunded -> [other] (illegal: skips) ❌
paid -> refunded ✅
cancelled -> refunded ✅
refunded_partially -> refunded ✅

**Fixed issue**:  Resolves #157 <!-- #-prefixed issue number -->

msilvagarcia reviewed (commented):
- The `currencyUtil` property in the `NotificationProcessor` class is not actually used.
- Why change the return value of `NotificationProcessorFactory#create` from `NotificationProcessorInterface` to `NotificationProcessor`?
- Should we worry about refund values?
-

msilvagarcia reviewed (changes requested):
https://github.com/Adyen/adyen-shopware6/pull/161#pullrequestreview-637412232

peterojo commented on src/NotificationProcessor/RefundNotificationProcessor.php:
```diff
@@ -0,0 +1,84 @@
+
+        if ($this->getNotification()->isSuccess()) {
+            $refundedAmount = (int) $this->getNotification()->getAmountValue();
+            $transactionAmount = $this->currencyUtil->sanitize(
```
@msilvagarcia currencyUtil is used here

peterojo commented:
@msilvagarcia 
> I love the refactor in the Order Repository. Great implementation :)
🍻
> * The `currencyUtil` property in the `NotificationProcessor` class is not actually used. Maybe that was something from another change, or something you planned to add and changed your mind? :)
it's used in the refund processor
> * Why change the return value of `NotificationProcessorFactory#create` from `NotificationProcessorInterface` to `NotificationProcessor`? Seems better to me to continue using the interface and update it accordingly if we need to add more methods, specially since we have multiple types of Notification Processors.
Good point, I should have left it alone, and implemented the interface on the abstract class `NotificationProcessor`, that's what was missing
> * Should we worry about refund values? How often does it occur that we receive a notification for a refund amount higher than the transaction amount?
The way I see it, we only care if the refund value is less because then it is a partial refund (Adyen just sends the same `REFUND` notification). If for some reason the refund amount is somehow greater than the transaction amount then it still counts as a full refund. What do you think @acampos1916

cyattilakiss commented:
> The way I see it, we only care if the refund value is less because then it is a partial refund (Adyen just sends the same REFUND notification). If for some reason the refund amount is somehow greater than the transaction amount then it still counts as a full refund. What do you think @acampos1916

@peterojo do we have a status like "needs attention" or anything like that? In case the amount is greater than the original one then it's definitely a potential issue. If possible I would flag it somehow while setting the status to refunded but indeed it should not happen.
If the amount is less then it's a partial refund indeed.

peterojo commented:
> @peterojo do we have a status like "needs attention" or anything like that? In case the amount is greater than the original one then it's definitely a potential issue. If possible I would flag it somehow while setting the status to refunded but indeed it should not happen.
> If the amount is less then it's a partial refund indeed.
Since we don't have such a status like 'needs attention', if the refunded amount is greater, then I will leave a warning log and *not* set the status to refunded. So that the merchant can be aware.

cyattilakiss commented on src/NotificationProcessor/RefundNotificationProcessor.php:
```diff
@@ -0,0 +1,90 @@
+                $this->getOrder()->getCurrency()->getIsoCode()
+            );
+
+            if ($refundedAmount > $transactionAmount) {
```
Just realised maybe it's also a good idea to add this as an error in the notification record error_message field therefore it's a bit more visible for the merchants what is the problem and with which notification

Action for the merchant can be the following:
Check whether the refund is indeed for the order indicated in the merchantReference. If yes then manually move the order to refunded and check why the value difference

If it's not for that order then it's good that you don't refund it in shopware and then the merchant can reach out to support to figure out why the wrong connection

peterojo replied:
that's a good point, in that case my solution would be to throw an exception here and it will be saved in the [handler](https://github.com/Adyen/adyen-shopware6/blob/df5929d343c71ae92adf751c190ea4d47c29d95f/src/ScheduledTask/ProcessNotificationsHandler.php#L138) as an error on the notification (also logged as an error).

cyattilakiss replied:
perfect :)

acampos1916 commented on src/NotificationProcessor/RefundNotificationProcessor.php:
```diff
@@ -0,0 +1,90 @@
+                $this->getTransactionStateHandler()->refund($orderTransaction->getId(), $context);
+            }
+        } catch (IllegalTransitionException $exception) {
+            // set to paid, and then try again
```
Maybe we could add a log here to let the merchant know that the transaction was moved to paid before being moved to refunded
